streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -570,6 +581,49 @@ void runLoop() {
         }
     }
 
+    public void setStreamsUncaughtExceptionHandler(final StreamsUncaughtExceptionHandler streamsUncaughtExceptionHandler, final KafkaStreams kafkaStreams) {
+        this.streamsUncaughtExceptionHandler = streamsUncaughtExceptionHandler;
+        this.kafkaStreams = kafkaStreams;
+    }
+
+    private void handleStreamsUncaughtException(final Exception e) {
+        final StreamsUncaughtExceptionHandler.StreamsUncaughtExceptionHandlerResponse action = this.streamsUncaughtExceptionHandler.handle(e);
+        if (kafkaStreams == null) {
+            log.error("Encountered the following exception during processing " +
+                    "and the thread is going to shut down: ", e);
+            return;
+        }
+        switch (action) {
+            case SHUTDOWN_STREAM_THREAD:
+                log.error("Encountered the following exception during processing " +
+                        "and the thread is going to shut down: ", e);
+                break;
+            case REPLACE_STREAM_THREAD:
+                log.error("Encountered the following exception during processing " +
+                        "and the the stream thread will be replaced: ", e);
+                break;
+            case SHUTDOWN_KAFKA_STREAMS_CLIENT:
+                log.error("Encountered the following exception during processing " +
+                        "and the client is going to shut down: ", e);
+                kafkaStreams.close(Duration.ZERO);
```

Comment:
currently leaves client state in not running instead of error. We need to make a close that leave in error and in non blocking